### ISSUE TITLE
react-compose context typings breaks Typescript 3.2.x

### ISSUE THREAD

**Issue by BPScott** [Shopify/quilt#489]
## Overview

When attempting to use typescript 3.2 we got lots of context warnings errors as components seem to expect explicit specification of props that provided by context (and thus you don't actually need to state them explicitly)

From https://shopify.slack.com/archives/CCNRSKLTH/p1545081241166000 (via @henrytao-me) and potentially related to https://shopify.slack.com/archives/CCNRSKLTH/p1546468712025100 (via @patsissons):

```ts
const Component = compose<Props>(
  fooEnhancer,
  barEnhancer,
)(StatelessComponent);

const result = mount(<Component baz="baz" />);
```

>The output `Component` from compose should not require `bar` and `foo` though.
The problem comes from all of the `C` generic here https://github.com/Shopify/quilt/blob/master/packages/react-compose/src/index.ts#L14. Removing all `C` works as expected.

I believe in this example `StatelessComponent` accepts the Prop `baz`, while `fooEnhancer` and `barEnhancer` provide the `foo` and `bar` props via context.

My typescript-fu isn't strong enough adequately explain why removing the `& C` fixes this or what knock on effects it may have.

## Consuming repo

> What repo were you working in when this issue occurred?

Within polaris-react this exposes itself on the [typescript 3.2](https://github.com/Shopify/polaris-react/tree/typescript-3.2) branch. Checkout that branch then run `yarn run sk type-check` and see several errors, that mostly relate to this issue - where it complains about a lack of `polaris` and `context` props that are provided via context.

**Comment by BPScott** [comment]
This also means that consumers of polaris can't update to TS 3.2: https://github.com/Shopify/polaris-react/issues/711

**Comment by TheMallen** [comment]
I think the `& C` is to capture any static methods / properties the component has outside of those that are on every `ReactComponent`.  Without it if your component has any subcomponents on it as static properties (some Polaris components do this iirc) TS would not know about them.

**Comment by patsissons** [comment]
It might be possible to shuffle `& C` to `hoistStatics` typing and eliminate the issue. That is how `recompose` does it (I think). It depends on where the type is getting inferred I suppose.

https://github.com/acdlite/recompose/blob/master/docs/API.md#hoiststatics

**Comment by BPScott** [comment]
[Continuing disclaimers that this is beyond my TS ability)

@lemonmade scribbled this up: https://github.com/Shopify/polaris-react/pull/803#issuecomment-452097651 as a way to grab just the static types of decorators. Might that help with the the potential for losing static methods/properties problem?

**Comment by TheMallen** [comment]
yeah that might work actually

**Comment by lemonmade** [comment]
lol @TheMallen don't sound so surprised :P

**Comment by patsissons** [comment]
Wow that is impressive!

**Comment by BPScott** [comment]
Heya folks, I've had two people open issues in polaris-react saying that typescript fails to compile and I believe the root cause is this issue. As time goes on more people are going to want to use TypeScript >= 3.2.0 and this issue is blocking them.

Would it be possible to take a look at this at some point?

**Comment by patsissons** [comment]
I think `react-compose` might become obsolete once we shuffle onto hooks

**Comment by henrytao-me** [comment]
@BPScott, you can try this wrapper for now if you use latest typescript. 

```typescript
import reactCompose, { WrappingFunction } from '@shopify/react-compose';
import React from 'react';

/**
 * Ref on how to fix static props / methods issue in Typescript
 * Issue: https://github.com/Shopify/quilt/issues/489
 * Solution by @lemonmade https://github.com/Shopify/polaris-react/pull/803#issuecomment-452097651
 */
export type StaticFields<Object> = {
  [Key in keyof Object]: Key extends 'prototype' ? never : Key
}[keyof Object];
export type Statics<Object> = Pick<Object, StaticFields<Object>>;

/**
 * This is a hack for @shopify/react-compose until this bug is fixed https://github.com/Shopify/quilt/issues/489
 */
export function compose<Props>(...wrappingFunctions: WrappingFunction[]) {
  return function wrapComponent<ComposedProps, C>(
    WrappedComponent: React.ComponentType<ComposedProps> & C
  ): React.ComponentType<Props> & Statics<typeof WrappedComponent> {
    return reactCompose(...wrappingFunctions)(WrappedComponent) as any;
  };
}
```

**Comment by BPScott** [comment]
> I think react-compose might become obsolete once we shuffle onto hooks

I agree and I'm excited for that future. However the timeline for Polaris bumping its minimum required version of react to 16.8 and then us updating all our components that currently use react-compose to use hooks and useContext() instead onto hooks is a looong time. If possible I'd like to not block every consumer of Polaris from updating past TypeScript 3.1.3 in the mean time

**Comment by lemonmade** [comment]
I get that everyone is busy, but this is the situation "PRs welcome" was designed for. The web team hasn't had a chance to get to this, because there are only 5–6 of us around, we have many other responsibilities, and Web does not suffer from this issue yet. I proposed a solution that works, and we'll be happy to review a PR from someone who has a bit of time to actually commit it.

**Comment by henrytao-me** [comment]
I can send a PR sometime today or tomorrow without your solution Chris.

**Comment by BPScott** [comment]
Fixed in #602
